### PR TITLE
8338737: Shenandoah: Reset marking bitmaps after the cycle

### PR DESCRIPTION
**Pull request by pengxiaolong**
WIP

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 1]
```diff
@@ -1187,6 +1198,23 @@ void ShenandoahConcurrentGC::op_cleanup_complete() {
...
+  ShenandoahWorkerScope scope(ShenandoahHeap::heap()->workers(),
+                          ShenandoahWorkerPolicy::calc_workers_for_conc_reset(),
+                          "reset after collection.");
+
+  ShenandoahHeap* const heap = ShenandoahHeap::heap();
+  if (heap->mode()->is_generational()) {
+    if (!_do_old_gc_bootstrap && !heap->is_concurrent_old_mark_in_progress()) {
+      // Only reset for young generation, bitmap for old generation must be retained,
+      // except there is collection(global/old/degen/full) trigged to collect regions in old gen.
+      heap->young_generation()->reset_mark_bitmap<false>();
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: Shouldn't it be safe to reset young region bitmaps even when old marking is in progress?

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp [hunk 2]
```diff
@@ -41,53 +41,44 @@
...
-    if (r->is_active()) {
-      // Reset live data and set TAMS optimistically. We would recheck these under the pause
-      // anyway to capture any updates that happened since now.
-      _ctx->capture_top_at_mark_start(r);
-      r->clear_live_data();
+  void heap_region_do(ShenandoahHeapRegion* region) override {
+    assert(!_heap->is_uncommit_in_progress(), "Cannot uncommit bitmaps while resetting them.");
+    if (PREPARE_FOR_CURRENT_CYCLE) {
+      if (region->need_bitmap_reset() && _heap->is_bitmap_slice_committed(region)) {
+        _ctx->clear_bitmap(region);
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:58]: Should this also `region->unset_need_bitmap_reset()`?

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp [hunk 3]
```diff
@@ -41,53 +41,44 @@
...
+      } else {
+        region->set_need_bitmap_reset();
+      }
+      // Capture Top At Mark Start for this generation.
+      if (region->is_active()) {
+        // Reset live data and set TAMS optimistically. We would recheck these under the pause
+        // anyway to capture any updates that happened since now.
+        _ctx->capture_top_at_mark_start(region);
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:66]: Full GC used to do this unconditionally for all affiliated regions. Do we not still need that to happen?

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp [hunk 4]
```diff
@@ -89,6 +89,7 @@ ShenandoahHeapRegion::ShenandoahHeapRegion(HeapWord* start, size_t index, bool c
     SpaceMangler::mangle_region(MemRegion(_bottom, _end));
   }
   _recycling.unset();
+  _need_bitmap_reset = true;
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:92]: Move to initializers? Why does it start with `true`? A new region would have a clear bitmap, right?

File: src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp [hunk 5]
```diff
@@ -157,5 +157,8 @@ bool ShenandoahOldGC::collect(GCCause::Cause cause) {
     LogStream ls(lt);
     result.print_on("Old Mark", &ls);
   }
+
+  entry_reset_after_collect();
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp:161]: Not sure we want to reset old region bitmaps after old marking is compete. Shenandoah opportunistically uses the bitmap for old regions during remembered set scan (it's faster than walking the heap).

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp [hunk 6]
```diff
@@ -266,6 +266,8 @@ class ShenandoahHeapRegion {
 
   ShenandoahSharedFlag _recycling; // Used to indicate that the region is being recycled; see try_recycle*().
 
+  bool _need_bitmap_reset;
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp:269]: Nit pick, but I think this would read better as `_needs_bitmap_reset`.

**Review by earthling-amzn — CHANGES_REQUESTED** [review]
Looks good. Left a few nits and a few questions in the review.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: Not really, if old marking is in progress, but current cycle is not bootstrap cycle, it means previous old collection has been cancelled to deal with allocation failure, control thread will try to resume old collection agin which will resume old marking again.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 3 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:66]: I will double check this, I'm not 100% sure

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 2 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:58]: It is for current cycles, bit will be dirty anyway, for next cycle reset will be needed, so the flag can't be unset here

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: The old cycle may be preempted by young collections, but it is only really _cancelled_ by global cycles or full GCs. Control thread will resume old marking, but this operates independently from young bitmap regions. I think we can reset young region bitmaps even when concurrent old marking is on going.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 2 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:58]: Got it.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp (hunk 5 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp:161]: The 'reset-after' won't reset bitmaps for old gen, it only reset for young gen, which is skipped in bootstrap cycle

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: I think we are taking the same thing, old gen could be preempted by young gc and resumed after the cycle. I have seem crash from caused by this, an old gc was bootstrapped but it was preempted/canceled multiple times right after the old gc started, eventually caused a crash from verifier because it expected the object in young is marked. I will share the gc log on slack later.

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp (hunk 6 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp]: Done.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 3 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:66]: I have update code to handle old GC differently, now the behavior should be same as before.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: That sounds like an issue with the verifier then? Once a young cycle is complete, nothing should depend on the state of the bitmaps for young regions (if, for no other reason, evacuation could have moved objects so that the bitmaps no longer represent the addresses of marked objects that were evacuated).

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 7]
```diff
@@ -236,6 +236,11 @@ bool ShenandoahConcurrentGC::collect(GCCause::Cause cause) {
   if (heap->mode()->is_generational()) {
     ShenandoahGenerationalHeap::heap()->complete_concurrent_cycle();
   }
+
+  // Instead of always reset before collect, some reset can be done after collect to save
+  // the time before before the cycle so the cycle can be started as soon as possible.
+  entry_reset_after_collect();
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:242]: For comment, I would say: "Instead of always resetting immediately before the start of a new GC, we can often reset at the end of the previous GC.  This allows us to start the next GC cycle more quickly after a trigger condition is detected, reducing the likelihood that GC will degenerate."

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 8]
```diff
@@ -578,12 +572,29 @@ void ShenandoahConcurrentGC::entry_cleanup_complete() {
...
+
 void ShenandoahConcurrentGC::op_reset() {
   ShenandoahHeap* const heap = ShenandoahHeap::heap();
   if (ShenandoahPacing) {
     heap->pacer()->setup_for_reset();
   }
-  _generation->prepare_gc();
+  // If it is old GC bootstrap cycle, always clear bitmap for global gen
+  // to ensure bitmap for old gen is clear for old GC cycle after this.
+  if (_do_old_gc_bootstrap) {
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:592]: This may deserve a comment.  It seems we ought to clear the old-gen mark bitmap at the end of coalesce-and-fill.  But that does not allow us to avoid clearing old-gen mark bitmaps at start of bootstrap because when young-gen regions are promoted in place, the mark bitmap is preserved for those regions, and since they are considered old at the end of the GC cycle during which they were promoted, those bitmaps will not be cleared by op_reset_after_collect().  Is there a way to improve this behavior?

For example, in op_reset_after_collect(), maybe we should clear old-gen bitmaps also (at least for recently promoted in place regions) unless old marking is in process and/or mixed evacuations are in progress.

Maybe this can be tackled in a separate PR, but would be good to file JBS ticket now if there is agreement on the approach.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: I agree with @earthling-amzn that we should be able to reset young-generation mark bitmap even if this is old_gc_bootstrap and even if old marking is in progress.  We should dive deeper to figure out the crash you observed.  It seems we don't fully understand the root cause.

I also suggest rewording the comment.  trigged?  (See other comments about increasing generality of this approach.)

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 8 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:592]: Yes, We can reset bimap of old region when there in place promotion and all old regions after coalesce-and-fill for old gen. 

Thanks Kelvin, I'll create a JBS ticket for this.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:1211]: I have tested it after removing `if (!_do_old_gc_bootstrap && !heap->is_concurrent_old_mark_in_progress())`, and always get crash in stress test like:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/codebuild/output/src48/src/s3/00/src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp:1270), pid=1578, tid=1595
#  Error: Remembered set violation at init-update-references; clean card should be dirty
Referenced from:
  interior location: 0x00000007f8000008
  inside Java heap
    not in collection set
  region: | 2528|R  |O|BTE    7f8000000,    7f8400000,    7f8400000|TAMS    7f8400000|UWM    7f8400000|U  4096K|T     0B|G  4096K|P     0B|S     0B|L   672B|CP   0
Object:
  0x00000007f5dc8b58 - klass 0x0000078000249400 java.lang.invoke.MethodType
    not allocated after mark start
    not after update watermark
        marked strong
    not marked weak
    not in collection set
  age: 8
  mark: mark(is_unlocked no_hash age=8)
  region: | 2519|R  |Y|BTE    7f5c00000,    7f6000000,    7f6000000|TAMS    7f6000000|UWM    7f6000000|U  4096K|T     0B|G  4096K|P     0B|S     0B|L  4091K|CP   0
```
It could be something wrong in remembered set scan, resetting young region bitmaps somehow tickles the issue.

I have created another [JBS ticket](https://bugs.openjdk.org/browse/JDK-8347371) to track the issue in remembered set scan, and keep this test for now. I'll update the comments in code.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 7 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:243]: Fixed. thanks!

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 8 above)
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:593]: This is ShenandoahConcurrentGC::op_reset(), it is executed when a cycle starts. I have removed line 361 to 371, which traverse all regions and apply reset for old regions when `_do_old_gc_bootstrap` is true, so it used to iterate regions twice when `_do_old_gc_bootstrap` is true. With this change, it only iterate once and reset bitmap for all regions when `_do_old_gc_bootstrap` is true.

Here is the ticket https://bugs.openjdk.org/browse/JDK-8347372 to follow up the possible improvements on old GC.

**Review by earthling-amzn — APPROVED** [review]
Okay, these changes look good. We don't yet understand why we cannot reset young bitmaps during an old cycle, but we will follow up that investigation separately.

**Comment by pengxiaolong** [comment]
Thanks!

/integrate
